update-format-crau/src/delta_update.rs

```diff
@@ -85,10 +92,64 @@ pub fn get_signatures_bytes<'a>(mut f: &'a File, header: &'a DeltaUpdateFileHead
     Ok(signatures_bytes.unwrap())
 }
 
+// Take a buffer reader, delta file header, manifest as input.
+// Return the whole data blobs, including header, manifest.
+pub fn get_header_data<'a>(f: &'a mut BufReader<File>, header: &'a DeltaUpdateFileHeader, manifest: &proto::DeltaArchiveManifest) -> Result<Vec<u8>, Box<dyn Error>> {
+    // Reset to the beginning of the stream, which means the whole buffer including
+    // delta update header as well as manifest. That is because data that must be verified
+    // with signatures start from the beginning. No header.translate_offset(0).
+    //
+    // Payload data structure:
+    //  | header | manifest | data blobs | signatures |
+
+    f.seek(SeekFrom::Start(0))?;
+
+    let data_length = header.translate_offset(manifest.signatures_offset.unwrap());
+
+    let mut header_data = vec![0u8; data_length as usize];
+    f.read_exact(&mut header_data)?;
+
+    Ok(header_data)
+}
+
+// Take a buffer reader, delta file header, manifest as input.
+// Return the data blobs, without header, manifest, or signatures.
+pub fn get_data_blobs<'a>(f: &'a mut BufReader<File>, header: &'a DeltaUpdateFileHeader, manifest: &proto::DeltaArchiveManifest) -> Result<Vec<u8>, Box<dyn Error>> {
+    // Read from the beginning of header, which means buffer including only data blobs.
+    // It means it is necessary to call header.translate_offset(), in contrast to
+    // get_header_data.
+    // Iterate each partition_operations to get data offset and data length.
+
+    let mut data_blobs = Vec::new();
```

Comment:
This extracts to RAM. When we want to deliver the dev container as sysext this will result in an OOM for many VMs.

Reply:
Changed data types from buf to file in several places.
That should resolve most OOM issues.

However, the code still has a place where simple read from payload into a memory buffer. Immediately before checking for RSA signature.
Main limitation here is that the RSA crate does not support streaming-based read/write. It takes only a simple buffer. So currently I have no better idea.

Or another alternative is to get away from the full-blown approach of RSA signatures like taking the whole buffer, and instead simply do not check the payload itself, just like the current update_engine. If we go down to that route, we cannot rely on the pure RSA, but we have to bring back openssl.

Reply:
Thanks. So, the RSA verification is done for the whole data instead of for a hash?

Reply:
Ah, I see `verify_digest` which is used internally to verify a hash.

Reply:
The signature is for the [sha256 hash](https://github.com/flatcar/update_engine/blob/d262b46e4c0595e6cefa726a19b8b8c1cdf167a5/decode_payload#L148), it should be possible to only pass the hash in, not the full data.

Reply:
We should use https://docs.rs/signature/2.1.0/signature/trait.DigestVerifier.html#tymethod.verify_digest or https://docs.rs/signature/2.1.0/signature/hazmat/trait.PrehashVerifier.html#tymethod.verify_prehash (both are like the simple `verify` exposed by https://docs.rs/rsa/latest/rsa/pkcs1v15/struct.VerifyingKey.html)

Reply:
Right. I have been trying to convert verify to verify_digest, to pass digest instead, but that change resulted in other build errors.

Reply:
Done with converting the verify function to prehash. Could not make digest work.